packages/next/src/server/app-render/parse-request-headers.ts

```diff
@@ -0,0 +1,56 @@
+import type { IncomingHttpHeaders } from 'http'
+import {
+  NEXT_ROUTER_PREFETCH_HEADER,
+  NEXT_ROUTER_STATE_TREE,
+  RSC_HEADER,
+} from '../../client/components/app-router-headers'
+import { getScriptNonceFromHeader } from './get-script-nonce-from-header'
+import { parseAndValidateFlightRouterState } from './parse-and-validate-flight-router-state'
+import type { FlightRouterState } from './types'
+
+export interface ParseRequestHeadersOptions {
+  readonly isRoutePPREnabled: boolean
+}
+
+export interface ParsedRequestHeaders {
+  /**
+   * Router state provided from the client-side router. Used to handle rendering
+   * from the common layout down. This value will be undefined if the request is
+   * not a client-side navigation request, or if the request is a prefetch
+   * request.
+   */
+  readonly flightRouterState: FlightRouterState | undefined
+  readonly isPrefetchRequest: boolean
+  readonly isRSCRequest: boolean
+  readonly nonce: string | undefined
+}
+
+// We read these values from the request object as, in certain cases,
+// base-server will strip them to opt into different rendering behavior.
+export function parseRequestHeaders(
```

Comment:
I'm not sure I see the value in this being a shared utility as right now as the return shape seems to be a grab-bag of properties that are currently only relevant to `renderToHTMLOrFlightImpl`. Maybe we should just move this function to be inside of `app-render` and not export it until we identify that this is relevant in more than 1 case to justify it being a shared util?

Although my 2c is that I don't feel like it warrants a separate function, and we could just move the `const csp = ...` part to be next to the other spot where we read `req.headers` for other things.

Reply:
This is indeed not the case of a shared utility. To me, there are more reasons than code re-use for extracting a piece of logic into a separate function/module. We also already have a couple of other cases like that in `app-render`, e.g. `createServerInsertedHTML` or `getRequiredScripts`.  In this case, the main purpose is to reduce the clutter in the main function, and it's done in preparation for #67527 where I add the parsing of another header.

But I'm also fine with reverting that, and maybe even reading the `Next-Fast-Refresh` header directly in the `RequestStore`, which I did consider originally, but discarded because it felt even more scattered.

Reply:
I'm not opposed to pulling it into a separate, non-exported function to improve readability - `app-render` has a handful of functions in there that aren't used outside of it (ie `generateFlight` or `createFlightDataResolver`). My minor concern with it being moved into a separate module with a name like `parseRequestHeaders` is that it suggests that _this_ particular util is the way to parse request headers, but that's not entirely the case either. For example, `base-server` pulls the data off of `requestMeta`:
https://github.com/vercel/next.js/blob/5df4e7e83af3039823c560c9e7dab5066efacfce/packages/next/src/server/base-server.ts#L1937-L1946

Reply:
Yeah that makes sense, though keep in mind that it's placed in `packages/next/src/server/app-render/`. But you're right, the function name sounds probably too generic. With auto-import you might end up in the situation you have in mind.

So the options are 1) inlining, 2) renaming (e.g. `getAppRenderContextFromHeaders` or something), or 3) discarding altogether (and reading the header in `withRequestStore`). Before moving forward, I think you need to take a look at #67527 because there the handling is changed a bit, and `parsedRequestHeaders` are added to `AppRenderBaseContext`.

Reply:
I vote for 1 -- I don't really mind when everything relevant to the function I'm looking at is in the same file; it requires less context switching and makes it immediately obvious it's not used by anything else. But we can wait for others to weigh in here as well.

Reply:
@wyattjoh Do you want to chime in here? Otherwise I would just go with option 1 then.

Reply:
For this case, I think to address the concerns of isolation and preventing creep, placing this function into the `app-render.tsx` as an unexported function would be best for now. Until we get more comprehensive ways to describe the different ways a given request can be reused, this keeps the isolation we want while keeping it internal to the `app-render` module.

I like the separate function over the inlining because it gives concrete types that can be described here, allowing good points for us to add comments/documentation to their uses and applications.